crates/biome_grit_patterns/src/pattern_compiler/snippet_compiler.rs

```diff
@@ -1,6 +1,191 @@
-use grit_util::{ByteRange, Language};
+use super::compilation_context::NodeCompilationContext;
+use crate::{
+    grit_code_snippet::GritCodeSnippet, grit_context::GritQueryContext,
+    grit_target_node::GritTargetNode, grit_tree::GritTree, CompileError,
+};
+use grit_pattern_matcher::{
+    constants::GLOBAL_VARS_SCOPE_INDEX,
+    pattern::{DynamicPattern, DynamicSnippet, DynamicSnippetPart, Pattern, Variable},
+};
+use grit_util::{Ast, AstNode, ByteRange, Language, SnippetTree};
 use std::borrow::Cow;
 
+pub(crate) fn parse_snippet_content(
+    source: &str,
+    range: ByteRange,
+    context: &mut NodeCompilationContext,
+    is_rhs: bool,
+) -> Result<Pattern<GritQueryContext>, CompileError> {
+    // we check for BRACKET_VAR_REGEX in the content, and if found
+    // compile into a DynamicPattern, rather than a CodeSnippet.
+    // This is because the syntax should only ever be necessary
+    // when treating a metavariable as a string to substitute
+    // rather than an AST node to match on. eg. in the following
+    // `const ${name}Handler = useCallback(async () => $body, []);`
+    // $name does not correspond to a node, but rather prepends a
+    // string to "Handler", which will together combine into an
+    // identifier.
+    if context
+        .compilation
+        .lang
+        .metavariable_bracket_regex()
+        .is_match(source)
+    {
+        return if is_rhs {
+            Ok(Pattern::Dynamic(
+                dynamic_snippet_from_source(source, range, context).map(DynamicPattern::Snippet)?,
+            ))
+        } else {
+            Err(CompileError::InvalidBracketedMetavariable)
+        };
+    }
+
+    if context
+        .compilation
+        .lang
+        .exact_variable_regex()
+        .is_match(source.trim())
+    {
+        return match source.trim() {
+            "$_" => Ok(Pattern::Underscore),
+            "^_" => Ok(Pattern::Underscore),
+            name => {
+                let var = context.register_variable(name.to_owned(), range)?;
+                Ok(Pattern::Variable(var))
+            }
+        };
+    }
+
+    let snippet_trees = context.compilation.lang.parse_snippet_contexts(source);
+    let snippet_nodes = nodes_from_indices(&snippet_trees);
+    if snippet_nodes.is_empty() {
+        // not checking if is_rhs. So could potentially
+        // be harder to find bugs where we expect the pattern
+        // to parse. unfortunately got rid of check to support
+        // passing non-node snippets as args.
+        return Ok(Pattern::Dynamic(
+            dynamic_snippet_from_source(source, range, context).map(DynamicPattern::Snippet)?,
+        ));
+    }
+
+    let dynamic_snippet = dynamic_snippet_from_source(source, range, context)
+        .map_or(None, |s| Some(DynamicPattern::Snippet(s)));
+    Ok(Pattern::CodeSnippet(GritCodeSnippet {
+        dynamic_snippet,
+        source: source.to_owned(),
```

Comment:
I noticed a lot of `to_owned` in this PR. Internally, we prefer to use `to_string` whenever possible unless it's a type that needs to use this function.